tests/test_fft.f90

```diff
@@ -0,0 +1,164 @@
+  real(dp), parameter :: tol=1e-10
+  logical :: use_2decomp
+
+#ifdef CUDA
+  type(cuda_backend_t), target :: cuda_backend
+  type(cuda_allocator_t), target :: cuda_allocator
+  integer :: ndevs, devnum
+#else
+  type(omp_backend_t), target :: omp_backend
+  type(allocator_t), target :: omp_allocator
+#endif
+
+  ! Initialise variables and arrays
+  call MPI_Init(ierr)
+  call MPI_Comm_rank(MPI_COMM_WORLD, nrank, ierr)
+  call MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
+
+#ifdef CUDA
+  ierr = cudaGetDeviceCount(ndevs)
+  ierr = cudaSetDevice(mod(nrank, ndevs)) ! round-robin
+  ierr = cudaGetDevice(devnum)
+  use_2decomp = .false.
+#else
+  use_2decomp = .true.
+#endif
+
+  ! Global number of cells in each direction
+  dims_global = [64, 32, 128]
+
+  ! Global domain dimensions
+  L_global = [2*pi, 2*pi, 2*pi]
+
+  ! Domain decomposition in each direction
+  nproc_dir = [1, 1, nproc]
+
+  BC_x = ['periodic', 'periodic']
+  BC_y = ['periodic', 'periodic']
+  BC_z = ['periodic', 'periodic']
+
+  mesh = mesh_t(dims_global, nproc_dir, L_global, BC_x, BC_y, BC_z, use_2decomp=use_2decomp)
+
+#ifdef CUDA
+  cuda_allocator = cuda_allocator_t(mesh, SZ)
+  allocator => cuda_allocator
+  print *, 'CUDA allocator instantiated'
+
+  cuda_backend = cuda_backend_t(mesh, allocator)
+  backend => cuda_backend
+  print *, 'CUDA backend instantiated'
+#else
+  omp_allocator = allocator_t(mesh, SZ)
+  allocator => omp_allocator
+
+  omp_backend = omp_backend_t(mesh, allocator)
+  backend => omp_backend
+#endif
+
+  if (nrank == 0) print *, 'Parallel run with', nproc, 'ranks'
+
+  allocate (xdirps, ydirps, zdirps)
+  xdirps%dir = DIR_X
+  ydirps%dir = DIR_Y
+  zdirps%dir = DIR_Z
+  call allocate_tdsops(xdirps, backend, 'compact6', 'compact6', &
+                       'classic', 'compact6', mesh%grid%BCs)
+  call allocate_tdsops(ydirps, backend, 'compact6', 'compact6', &
+                       'classic', 'compact6', mesh%grid%BCs)
+  call allocate_tdsops(zdirps, backend, 'compact6', 'compact6', &
+                       'classic', 'compact6', mesh%grid%BCs)
+
+  input_field => allocator%get_block(DIR_C, CELL)
+  output_field => allocator%get_block(DIR_C, CELL)
+
+  dims = mesh%get_dims(CELL)
+  input_field%data(:, :, :) = 0.0
+  output_field%data(:, :, :) = 0.0
+
+
+  ! Initialise field with some function
+  do k = 1, dims(3)
+    do j = 1, dims(2)
+      do i = 1, dims(1)
+        xloc = mesh%get_coordinates(i, j, k)
+        x = xloc(1)
+        y = xloc(2)
+        z = xloc(3)
+        input_field%data(i, j, k) = sin(x)*cos(y)*cos(z) + 2*x
+      end do
+    end do
+  end do
+
+  call backend%init_poisson_fft(mesh, xdirps, ydirps, zdirps)
+
+  ! Compute FFT and back
+  call backend%poisson_fft%fft_forward(input_field)
+  call backend%poisson_fft%fft_backward(output_field)
+
+  ! Output scaled with number of cells in domain
+  error_norm = norm2(input_field%data(:, :, :) - output_field%data(:, :, :)/product(dims_global) )**2/product(dims_global)
```

Comment:
Just to double check - the first division by problem size is normalising the backwards FFT, i.e. intentional, and 2nd is part of RMS operation?

Reply:
yes exactly. Do you think it needs extra comments?

Reply:
you are right, just added one.